src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +452,317 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+            {
+                return;
+            }
+
+            var document = SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document is null)
+            {
+                return;
+            }
+
+            var newSymbolKey = (e.SelectedItem as AbstractSignatureHelpProvider.SymbolKeySignatureHelpItem)?.SymbolKey ?? default;
+            var newSymbol = newSymbolKey.Resolve(document.Project.GetRequiredCompilationAsync(CancellationToken.None).WaitAndGetResult(CancellationToken.None), cancellationToken: CancellationToken.None).GetAnySymbol();
+            if (newSymbol is not IMethodSymbol method)
+                return;
+
+            MoveToSymbol(_state._symbols, method, CancellationToken.None);
+        }
+
+        private async Task<ImmutableArray<ISymbol>> GetSymbolsAsync(
+            SnapshotPoint caretPosition,
+            CancellationToken cancellationToken)
+        {
+            var document = SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document is null)
+            {
+                return ImmutableArray<ISymbol>.Empty;
+            }
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var token = await semanticModel.SyntaxTree.GetTouchingTokenAsync(caretPosition.Position, cancellationToken).ConfigureAwait(false);
+            var semanticInfo = semanticModel.GetSemanticInfo(token, document.Project.Solution.Workspace, cancellationToken);
+            return semanticInfo.ReferencedSymbols;
+        }
+
+        public void MoveToSymbol(ImmutableArray<IMethodSymbol> symbols, IMethodSymbol symbol, CancellationToken cancellationToken)
+        {
+            AssertIsForeground();
+
+            if (ExpansionSession is null)
+            {
+                return;
+            }
+
+            if (SymbolEqualityComparer.Default.Equals(_state._symbol, symbol))
+            {
+                return;
+            }
+
+            var symbolName = _state._symbol?.Name ?? _state._symbols.FirstOrDefault()?.Name;
+            if (symbolName != symbol.Name)
+            {
+                // Can't track this request
+                _state.ClearSymbolInformation();
+                return;
+            }
+
+            if (!_state._symbols.IsDefaultOrEmpty && _state._symbols != symbols)
+            {
+                // Can't track this request
+                _state.ClearSymbolInformation();
+                return;
+            }
+
+            var textViewModel = TextView.TextViewModel;
+            if (textViewModel == null)
+            {
+                Debug.Assert(TextView.IsClosed);
+                return;
+            }
+
+            var buffer = EditorAdaptersFactoryService.GetBufferAdapter(textViewModel.DataBuffer);
+            if (buffer is not IVsExpansion expansion)
+            {
+                return;
+            }
```

Comment:
i'd love if all the UI touching/checking moved to the caller.

Reply:
➡️ Not moving to caller. This type is _heavily_ UI thread bound with no intent to change outside of a rewritten platform API for snippets.